### DISCUSSION TITLE
Aspire Workload 8.2.0

### DISCUSSION THREAD

**Discussion by jdasilva-olo** [microsoft/aspire#5488] [Q&A] [answered]
I'm having an issue where I've got an Aspire build based on 8.1.0 packages, but now `dotnet workload install aspire` installs an 8.2.0 sdk which fails the build.  The error message is `TestAppHost.Redis is a .NET Aspire AppHost project that needs a package reference to Aspire.Hosting.AppHost version 8.2.0 or above to work correctly. You are using version 8.1.0.`

I'm using `dotnet-install.sh` to install the .NET SDK `-channel 8.0` locally and afaict the .NET SDK version did not change.
```
.NET SDK:
Version:           8.0.401
Commit:            811edcc344
Workload version:  8.0.400-manifests.b6724b7a
MSBuild version:   17.11.4+37eb419ad
```

I've tried everything I could find to force 8.1.0 to be installed.

`dotnet workload install aspire --version "8.1.0"` results in: 
`Workload installation failed: Failed to install workload version 8.1.0: Version 8.0.0 of package microsoft.net.workloads.8.1.0 is not found in NuGet feeds https://api.nuget.org/v3/index.json".`

Using `--from-rollback-file` with the content below results:
`No workloads installed for this feature band. To update workloads installed with earlier SDK versions, include the --from-previous-sdk option.` 
So I also tried with `--from-previous-sdk` to no avail.

Am I missing some way to force 8.1.0?  Or must I upgrade all of the packages to 8.2.0? 

[workload-rollback.json](https://github.com/user-attachments/files/16808648/workload-rollback.json)

**Comment by davidfowl** [comment]
Easier path is to move everything to 8.2.0

**Reply by gf-mbr** [reply]
I have the same issue where it now breaks my build pipeline.

If I force the version in the restore command, it works (`dotnet workload restore ./MySolution.sln --version 8.0.400`). However, I don't believe this is the correct place to force the version?

Trying to set the same workload version in the `global.json` results in:
`Unhandled exception: Microsoft.Build.Exceptions.InvalidProjectFileException: SDK Resolver Failure: "The SDK resolver "Microsoft.DotNet.MSBuildWorkloadSdkResolver" failed while attempting to resolve the SDK "Microsoft.NET.SDK.WorkloadAutoImportPropsLocator". Exception: "System.IO.FileNotFoundException: Workload version 8.0.400, which was specified in /home/runner/work/my-repo/my-repo/global.json, was not found. Run "dotnet workload restore" to install this workload version.`

What is the intended way to prevent these issues?

**Reply by Emil-Granbom** [reply]
We have a similar issue. We have updated all the packages to 8.2.0 only to get an error in VS:
error ASPIRE005: Advisor.Aspire.AppHost project requires a newer version of the .NET Aspire Workload to work correctly. Please run `dotnet workload update`.
Running dotnet workload update as suggested gives the following output
❯ dotnet workload update
Installing workload version 8.0.401.
Downloading Microsoft.NET.Sdk.Aspire.Manifest-8.0.100.Msi.x64 (8.2.0)
Removing Microsoft.NET.Sdk.Aspire.Manifest-8.0.100.Msi.x64 ..... Done

Successfully updated workload(s): android aspire ios maccatalyst maui-windows.

Running dotnet workload list reports
❯ dotnet workload list

Workload version: 8.0.401

Installed Workload Id      Installation Source
----------------------------------------------------------
android                    SDK 8.0.400, VS 17.11.35222.181
aspire                     SDK 8.0.400, VS 17.11.35222.181
ios                        SDK 8.0.400, VS 17.11.35222.181
maccatalyst                SDK 8.0.400, VS 17.11.35222.181
maui-windows               SDK 8.0.400, VS 17.11.35222.181

And still the VS builds fail with the same error ...

**Comment by davidfowl** [comment]
cc @joperezr

**Comment by joperezr** [comment, marked as answer]
Hello @jdasilva-olo, thank you for starting this discussion and I do want to apologize for the inconvenience this change has caused. I have posted [this breaking change issue](https://github.com/dotnet/aspire/issues/5501) with details and workarounds that should be able to unblock you if you want to pin to 8.1.0 or show you how to upgrade to 8.2.0. Please let me know if the workarounds in there don't work for you.

**Reply by jdasilva-olo** [reply]
Thank you @joperezr , that was very helpful and it did unblock me.  I've got an Aspire hosting package with common code for our org and I think it will be valuable to be able to provide updates without necessarily upping the version.  (It's still early in our usage though so I probably will choose to do the update next week anyway.)

One interesting thing I stumbled onto by accident was that even being on `workload-set` keeps the max version at `8.1.0` which made some sense to me given the context that same .NET SDK was being used and I didn't understand why Aspire seemed to "jump ahead".

Really appreciate the quick responses from the team, thank you!

I have not tried the newly added workaround yet using `SkipAddAspireDefaultReferences`.  Would the correct interpretation of that be, that I'm building against 8.1.0 packages but running against the 8.2.0 dashboard and orchestrator (say for a unit test)?

**Reply by joperezr** [reply]
> Would the correct interpretation of that be, that I'm building against 8.1.0 packages but running against the 8.2.0 dashboard and orchestrator (say for a unit test)?

That's exactly right, which was the behavior we had with previous workloads since some internal pieces of Aspire were coming from the workload. Starting with 8.2.0, all of these internal pieces will be brought in via packages and will match the version that you build against, which will help having more consistent behaviors across systems.

**Reply by aa-ko** [reply]
Hi everyone! I just found this discussion, because I had the same issue @jdasilva-olo described in their initial question. Updating to 8.2.0 fixed this for me, but I'm still wondering how this is possible in the first place.

My project references Aspire 8.0.2, resulting in my build breaking after the 8.2.0 release. This is the expected behavior, breaking changes are unavoidable.
However, I specifically pinned the version to 8.0.2, resulting in this error message when trying to install the workload:
`Workload update failed: Failed to install manifest microsoft.net.sdk.aspire version 8.0.2: Version 8.0.2 of package microsoft.net.sdk.aspire.manifest-8.0.300 is not found in NuGet feeds https://api.nuget.org/v3/index.json"..`

This makes no sense to me; version 8.0.2 is still visible on NuGet and no other dependency in my project changed. Yet, I cannot install the workload, resulting in my CI completely bricking over night.
@joperezr Could you please explain how this is possible?

**Reply by joperezr** [reply]
Can you share a bit more information on how you are trying to pin to 8.0.2? Pinning workloads (before workload-sets) isn't very straight forward, so I'd like to understand better how you are doing it so that I can help get you back to a working state.

**Reply by aa-ko** [reply]
Hi, sorry for the delay! My vacation got in the way :smile:

I was trying to both pin the SDK version and the Aspire workload version like this:

**global.json**
```json
{
  "sdk": {
    "rollForward": "disable",
    "allowPrerelease": false,
    "version": "8.0.401"
  }
}
```


**rollback.json**
```json
{
 "microsoft.net.sdk.aspire": "8.0.2"
}
```

The resulting log made me think that this worked fine, but the Aspire 8.0.2 workload could not be downloaded as if it had been removed from NuGet.

I fixed this by just updating to 8.2.0 in the mean time, so I'm not relying on this working, but I was just wondering. Thanks for making the effort to help!

**Reply by joperezr** [reply]
I see, your rollback.json file is not quite right which might be the reason why you are having issues pinning down. The advise is definitely to move to 8.2.0 (as you have done now) and also know that in the future (release 8.2.1) you will no longer be forced to move in order to use the new workload, but if you really wanted to pin down to 8.0.2 version the contents of the rollback should be:

```json
{
  "microsoft.net.sdk.aspire": "8.0.2/8.0.100"
}
```

That, plus running `dotnet workload update --from-rollback-file ./rollback.json` should do the trick.

**Reply by aa-ko** [reply]
That explains a lot! The error message suggested that Aspire 8.0.2 simply did not exist any more, hence my confusion. I will try to track the latest version anyways, so I have no need to pin to a specific version right now, but I'm sure knowing the correct syntax will be useful in the future.

Thank you for the clarification! :tada:
